### PR TITLE
Prerequisite for tracing chain building: refactor `find_issuer()` etc. in `x509_lu.c` and `x509_vfy.c`

### PR DESCRIPTION
**Pull request by DDvO**
This is a prerequisite for #18761. When preparing the new trace feature for cert chain building, I found again that there is an ugly overlap between
* `STACK_OF(X509) *X509_STORE_CTX_get1_certs(X509_STORE_CTX *ctx, const X509_NAME *nm)`
* `X509_STORE_CTX_get1_issuer()` in `x509_lu.c` used to search for suitable trusted certs
* the local function `find_issuer()` in `x509_vfy.c` used to search for both untrusted and trusted certs.

Among others, this means that any changes (such as the addition of trace output) would need to be done consistently in their function bodies.

Moreover, most of `X509_STORE_CTX_get1_issuer()` does not really belong to `x509_lu.c` but to `x509_vfy.c`,
but not all of its implementation can be moved because it includes collecting certs via the internal X509/CRL object lookup mechanism.
When adding the tracing for chain building, I also noticed that the function checks the first cert found twice, which of course is needless and should be avoided.

This PR fixes all this, while not changing the documented/commented semantics of these functions,
by letting `X509_STORE_CTX_get1_issuer()` make use of `X509_STORE_CTX_get1_certs()`
~by factoring out from `X509_STORE_CTX_get1_certs()` and `X509_STORE_CTX_get1_issuer()` a new function~

and moving the rest to `x509_vfy.c`.

These functions are now implemented using a more general internal function
```
static X509 *get0_best_issuer_sk(X509_STORE_CTX *ctx, int trusted,
                                 int no_dup, STACK_OF(X509) *sk, X509 *x)
```
which will be extended by #18761 to provide tracing.

This PR also renames the local function `get_issuer_sk()` to `get1_best_issuer_other_sk()` in `x509_vfy.c`.

### CODE REVIEW

File: crypto/x509/x509_vfy.c
Comment (kaduk) [inline crypto/x509/x509_vfy.c:134]: Do the nested ternarys really make the code more readable and maintainable?

File: crypto/x509/x509_vfy.c
Comment (kaduk) [inline crypto/x509/x509_vfy.c:3191]: I see that there are preexisting instances of this comment style (on the same line as the opening brace of the block) in the file (e.g., line 3192 in the "left" side), but I don't see it covered in https://www.openssl.org/policies/technical/coding-style.html#chapter-8-commenting and it's unclear that we want to be adding more of them.

**Review by kaduk — COMMENTED** [review]
A couple thoughts noticed while skimming to assess if I can do a "real" review in the time allotted (answer: no).

File: crypto/x509/x509_vfy.c
Comment (t8m) [reply crypto/x509/x509_vfy.c:134]: Yeah, I do not see them readable either - At least I would use parentheses around the internal ternary but IMO the original is more readable.

Comment (DDvO) [reply crypto/x509/x509_vfy.c:134]: In general, I like conditional expressions (using the `? :` ternary operator) for their conciseness.
Yet I agree that at least here the code is less readable because the conditions are nested.
I'm going to revert this type of changes.

File: crypto/x509/x509_vfy.c
Comment (DDvO) [reply crypto/x509/x509_vfy.c:3191]: The coding style guide says:
> Place comments above or to the right of the code they refer to.

so I see this covered.

Here I moved the part of the comment that refers to the condition to the right of the condition because this makes more clear what it refers to.

**Comment by DDvO** [comment]
Ping for reviews.

I wonder if anyone (except @kaduk, who cannot take the time) already plans or has started doing a review?

**Comment by DDvO** [comment]
Would it help if I split this into two PRs:
* one with all the just shallow coding style and doc improvements
* one with the actual 'meat', re-structuring the various  'find issuer' functions?

**Comment by t8m** [comment]
> Would it help if I split this into two PRs:
> 
>     * one with all the just shallow coding style and doc/comment improvements
> 
>     * one with the actual 'meat', re-structuring the various  'find issuer' functions

Not necessarily separate PRs but separate commits would probably help.

**Comment by DDvO** [comment]
> > Would it help if I split this into two PRs:
> > ```
> > * one with all the just shallow coding style and doc/comment improvements
> > 
> > * one with the actual 'meat', re-structuring the various  'find issuer' functions
> > ```
> 
> Not necessarily separate PRs but separate commits would probably help.

Done. 
And rebased the two commits, just in case.

File: crypto/x509/x509_lu.c
Comment (t8m) [inline crypto/x509/x509_lu.c:822]: Why changing ctx above and not here?

File: crypto/x509/x509_vfy.c
Comment (t8m) [inline crypto/x509/x509_vfy.c]: I'm not sure this is a better name for the parameter.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/x509/x509_lu.c
Comment (DDvO) [reply crypto/x509/x509_lu.c:810]: > why not `store`?

Simply because `cx` is short enough for avoiding the obligation to spilt the line due to the 80-chars width limit. Or how about using `s` here?

File: crypto/x509/x509_lu.c
Comment (DDvO) [reply crypto/x509/x509_lu.c:822]: > Why changing ctx above and not here?

Because here the line anyway had to be split, and after that there is enough space for the usual name `ctx`.

File: crypto/x509/x509_vfy.c
Comment (DDvO) [reply crypto/x509/x509_vfy.c]: I thought that the `p_` prefix may help to highlight that this is a double pointer,
and `res` makes clear that this is for the result.

File: crypto/x509/x509_lu.c
Comment (t8m) [reply crypto/x509/x509_lu.c:810]: The obligation is not hard obligation and reasonable variable/argument names are more important than this obligation.

Anyway `xs` would work for me and I'd use it also for other X509_STORE * arguments in other functions instead of ctx as X509_STORE * is not a context.

File: crypto/x509/x509_lu.c
Comment (t8m) [reply crypto/x509/x509_lu.c:822]: I would not split the line. The 80 characters limit is not a hard one.

File: crypto/x509/x509_vfy.c
Comment (t8m) [reply crypto/x509/x509_vfy.c]: I'd really prefer keeping the original. IMO it is clear enough.

File: crypto/x509/x509_lu.c
Comment (DDvO) [reply crypto/x509/x509_lu.c:822]: Nice, but then the nuisance is that `check-format.pl` will complain.

File: crypto/x509/x509_lu.c
Comment (DDvO) [reply crypto/x509/x509_lu.c]: Good point - changed all such occurrences, also in `x509_vfy.h.in` and in the respective docs.

File: crypto/x509/x509_lu.c
Comment (DDvO) [reply crypto/x509/x509_lu.c]: Anyway, reverted splitting the line here.

**Comment by DDvO** [comment]
I've just placed the new fixup commit in order,
such that future merge conflicts on autosquash will be avoided.

**Comment by t8m** [comment]
The codingstyle changes LGTM now. However the refactoring will need more time to review which I do not have right now.

**Comment by DDvO** [comment]
> The codingstyle changes LGTM now. However the refactoring will need more time to review which I do not have right now.

Thank you for letting me know.

Then I suppose it is best now if I copy the codingstyle commit(s) to a different PR, which you could approve right away, and after this is merged (likely soon), I'll rebase this PR and this PR will become simpler to review.
Ok?

**Comment by t8m** [comment]
OK

**Comment by DDvO** [comment]
Since #18840 has been merged now, I've been able to rebase this PR on it,
which got simplified accordingly.

So reviewing this PR can now continue.

**Comment by DDvO** [comment]
Can we please go on with this?
Both reviews are missing.

**Comment by DDvO** [comment]
Ping for reviews/approvals.
This is blocking #18761, which will add a trace facility to cert path construction.

**Comment by DDvO** [comment]
@openssl/committers can we please bring this forward.

As mentioned, this is blocking #18761, which will add a trace facility to cert path construction.

**Comment by DDvO** [comment]
@openssl/committers  can we please bring this forward?

This is still blocking #18761, which will add a very helpful feature: 
a trace facility for X.509 chain building (i.e., cert path construction,  the first step of certificate validation).

**Comment by DDvO** [comment]
I wonder why since 5 months nobody has taken up reviewing this PR further?

As mentioned, this is blocking an important enhancement: a trace facility for X.509 chain building,
which would be a great debugging aid on failures of cert path construction, the first step of certificate validation.

**Comment by t8m** [comment]
@DDvO I am really sorry for not getting to this yet. The certificate verification code requires very careful review which means I would need to spend at least one or two hours looking at the changes. I'll try to make this a priority somehow but I cannot promise anything for now.

**Comment by DDvO** [comment]
> @DDvO I am really sorry for not getting to this yet. The certificate verification code requires very careful review which means I would need to spend at least one or two hours looking at the changes. I'll try to make this a priority somehow but I cannot promise anything for now.

Thank you @t8m for letting me know.
This was about a month back - do you see any chance for doing the review within the next few weeks?

**Comment by DDvO** [comment]
Unfortunately, I missed for quite a while that this PR needed to be updated.
Done now.

@t8m, ready for (re-)review again.

**Comment by DDvO** [comment]
The CI **Coding style validation / check-style** failure is unrelated.

For some reason, its call  to`git rev-parse HEAD` gave a wrong (non-existing?) commit ID.
should be `d24ef3bf2de831b05f1461a7e66e07ad7a24fd8a`

**Review by t8m — DISMISSED** [review]

**Comment by t8m** [comment]
@openssl/committers let's get this reviewed.

**Comment by DDvO** [comment]
Rebased again to fix merge conflict.

Ping @openssl/committers for 2nd review

**Review by t8m — DISMISSED** [review]

**Review by hlandau — APPROVED** [review]

File: crypto/x509/x509_lu.c
Comment (vdukhovni) [inline crypto/x509/x509_lu.c:681]: Some day, it might be good to not assume that cached data is *always* complete and correct. Trust anchors can become untrusted, and new ones can come along. Certainly with CApath, it makes some sense to periodically ignore the cached data, and do a fresh lookup, perhaps a new control can set the refresh interval. Nothing needs to be done in this PR, but something to keep in mind...

File: crypto/x509/x509_lu.c
Comment (vdukhovni) [inline crypto/x509/x509_lu.c:659]: Does the `up_refs == 0` case really  make sense as an API?  Without the ref count bump the returned objects may be invalid by the time the caller looks at them. What is the use-case for `up_refs == 0`?

File: crypto/x509/x509_vfy.c
Comment (vdukhovni) [inline crypto/x509/x509_vfy.c:427]: With `up_refs == 0`, is this really thread safe?

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: crypto/x509/x509_lu.c
Comment (DDvO) [reply crypto/x509/x509_lu.c:679]: Good point.
I suggest opening an issue to help keeping this in mind.

File: crypto/x509/x509_lu.c
Comment (DDvO) [reply crypto/x509/x509_lu.c]: Oh, right, this is not thread-safe.
It was meant as minor optimisation, but not a good idea.
Reverted this to `X509_STORE_CTX_get1_certs()` only.

File: crypto/x509/x509_vfy.c
Comment (DDvO) [reply crypto/x509/x509_vfy.c]: Fixed (see above)

**Review by t8m — APPROVED** [review]
Approved if CI passes

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**
